lib/spack/spack/spec.py

```diff
@@ -1840,7 +1837,7 @@ def from_yaml(stream):
         stream -- string or file object to read from.
         """
         try:
-            data = syaml.load(stream)
+            data = yaml.load(stream)
```

Comment:
`spack_yaml` now refers to a yaml object specifically meant for parsing Spack configs. Applying it to yaml content generated from Specs can lead to errors. I'll add more in my response to:

> This probably needs an explanatory comment. Also, I'm not understanding the @ part -- what is that?

Reply:
Actually that only partially addresses this question. That establishes that we need special parsing to resolve an ambiguity with `:`. The issue is that when writing some specs not all specs with `:` contain `@`: the exception is `test_normal_spec` which ends up writing something like:

```
- mpi:
  versions:
  - ':'
  concrete: False
  hash: ...
```

So I decided that `spack_yaml` should only be used for Spack config files and regular yaml should be used for other cases (like turning Specs into yaml).

Reply:
Ok, sounds good.  I'll just go through the caveats to make sure I believe we're not relying on anything special in Spack YAML for this case:

1. Spack YAML uses `OrderedDicts` (`syaml_dict`) -- I don't *think* this matters for `spec.yaml`, as we go to great lengths to make sure it parses in any order but outputs in a deterministic order.  If these are the only two places we output to YAML, then I think this is fine.  Also, don't do it here, but we should probably eventually switch `spec.yaml` to be JSON just because it's faster.  The only reason it's still YAML is because we didn't want to change all the hashes.

2. Spack YAML has marks for file/line numbers.  I do not think that is used anywhere other than in config parsing, so I guess that is safe too.

Ok, I think this is a good idea.  I have on request, though, which is that I think the *Spack YAML* case should be special, not the normal YAML case.  So there should be `dump_config_yaml` or something, along with clear docs, NOT `dump_noconfig_yaml`.  I think the latter is confusing.

Reply:
> Spack YAML uses OrderedDicts (syaml_dict) -- I don't think this matters for spec.yaml, as we go to great lengths to make sure it parses in any order but outputs in a deterministic order. 

When constructing a dictionary to pass to `spack_yaml`, a `syaml_dict` is used which is ordered, so that will still be handled properly.